datahub/dnb_api/views.py

```
@@ -28,29 +30,62 @@ def post(self, request):
+        upstream_response = search_dnb(request)
         return HttpResponse(
             upstream_response.text,
             status=upstream_response.status_code,
             content_type=upstream_response.headers.get('content-type'),
         )
 
-    def _get_upstream_response(self, request):
 
-        if not settings.DNB_SERVICE_BASE_URL:
-            raise ImproperlyConfigured('The setting DNB_SERVICE_BASE_URL has not been set')
-        search_endpoint = urljoin(f'{settings.DNB_SERVICE_BASE_URL}/', 'companies/search/')
-        shared_key_auth = TokenAuth(settings.DNB_SERVICE_TOKEN)
-        api_client = APIClient(
-            search_endpoint,
-            shared_key_auth,
-            raise_for_status=False,
+class DNBCompanyCreateView(APIView):
+    """
+    View for creating datahub company from DNB data.
+    """
+
+    required_scopes = (Scope.internal_front_end,)
+    permission_classes = (IsAuthenticatedOrTokenHasScope,)
+
+    @method_decorator(feature_flagged_view(FEATURE_FLAG_DNB_COMPANY_SEARCH))
+    @method_decorator(enforce_request_content_type('application/json'))
+    def post(self, request):
+        """
+        Proxy for POST requests.
+        """
+        dnb_response = search_dnb(request).json()
+
+        # Query should return a single company
+        try:
+            dnb_company = dnb_response['results'].pop(0)
+        except IndexError:
+            error_message = 'Cannot find a company with duns_number: {duns_number}'
+            logger.error(error_message)
+            return JsonResponse(
+                {'error': error_message},
+                status=400,
+                content_type='application/json',
+            )
+        # If the dnb-service returns multiple companies for a duns_number
+        # we are going to create a company from the first one. This should
+        # not happen.
+        if not dnb_response['results']:
+            warning_message = 'Multiple companies returned for duns_number:{duns_number}'
```

I would be tempted to just fail early here with a 400 and meaningful error message rather than assuming that the user would want the first result.  Since there's currently no checking on the request body, somebody could call this with `{"search_term": "sainsbury", "address_country": "GB"}` and it would create a DH company for the first (of many) results.

---

My opinion might change if we had some input validation before this to assert that the user is only searching with `duns_number`.